### PR TITLE
Implement cn_nameiop equivalent

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**

### CODE REVIEW

**Comment by SuchAFuriousDeath** [comment]
@ultimaweapon This is roughly what I had in mind.

File: src/kernel/src/fs/path.rs [hunk 1]
```diff
@@ -286,11 +287,61 @@ impl From<VPathBuf> for String {
...
+    Lookup,
+    Create,
+    Delete,
+    Rename,
+}
+
+/// Inspired by [`std::path::Component`].
+/// Is not an equivalent of the FreeBSD type `componentname`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ComponentName<'a> {
```
Comment (ultimaweapon) [inline src/kernel/src/fs/path.rs:301]: `VPathComponent` should be a better name.

File: src/kernel/src/fs/mod.rs [hunk 2]
```diff
@@ -152,14 +152,25 @@ impl Fs {
...
+        &self,
+        path: impl AsRef<VPath>,
+        td: Option<&VThread>,
+    ) -> Result<Arc<Vnode>, LookupError> {
+        self.namei(path, NameiOp::Lookup, td)
+    }
+
     /// This method will **not** follow the last component if it is a mount point or a link.
-    pub fn lookup<P: AsRef<VPath>>(
+    fn namei(
```
Comment (ultimaweapon) [inline src/kernel/src/fs/mod.rs:170]: I'm not really a fan of `namei` because it make no sense.

File: src/kernel/src/fs/mod.rs [hunk 3]
```diff
@@ -152,14 +152,25 @@ impl Fs {
...
     }
 
+    pub fn lookup(
+        &self,
+        path: impl AsRef<VPath>,
+        td: Option<&VThread>,
+    ) -> Result<Arc<Vnode>, LookupError> {
+        self.namei(path, NameiOp::Lookup, td)
+    }
+
```
Comment (ultimaweapon) [inline src/kernel/src/fs/mod.rs:168]: We probably don't need this due to my previous comment.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/fs/path.rs [hunk 4]
```diff
@@ -286,11 +287,60 @@ impl From<VPathBuf> for String {
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum LookupOp {
```
Comment (ultimaweapon) [inline src/kernel/src/fs/path.rs:291]: This one should not be here.

File: src/kernel/src/fs/path.rs [hunk 5]
```diff
@@ -286,11 +287,60 @@ impl From<VPathBuf> for String {
...
+    Rename,
+}
+
+/// Inspired by [`std::path::Component`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum VPathComponent<'a> {
+    Root,
+    Dot,
+    DotDot,
+    Normal { name: &'a str },
```
Comment (ultimaweapon) [inline src/kernel/src/fs/path.rs:304]: I think `Name(&'a str)` looks better. It is also much easier to work with unnamed field than named field.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/fs/path.rs (hunk 4 above)
Comment (ultimaweapon) [reply src/kernel/src/fs/path.rs:291]: Seems like to missed this one. It should be in the root of `fs` instead of `path`.

**Pull request closed without merging**